src/tensor_ops/upscale2d/mod.rs

```diff
@@ -0,0 +1,210 @@
+                ),
+                E,
+                D,
+                T,
+            >;
+
+            fn try_upscale2d(self) -> Result<Self::Output, Self::Err> {
+                let &(chan, _, _) = self.shape();
+                let op = Upscale2DOp::new([1, chan.size(), H, W], [OH, OW]);
+                let (inp, mut tape) = self.split_tape();
+                let mut out =
+                    inp.device
+                        .try_zeros_like(&(chan, Default::default(), Default::default()))?;
+                inp.device.forward(op, &inp, &mut out)?;
+                let phantom_out = out.clone();
+                tape.try_alloc_grad(&inp)?;
+                tape.try_alloc_grad(&out)?;
+                tape.add_backward_op(move |grads| {
+                    let (grad_inp, grad_out) = grads.mut_and_ref(&inp, &phantom_out);
+                    inp.device
+                        .backward(op, &inp, grad_inp, &phantom_out, grad_out)
+                });
+                Ok(out.put_tape(tape))
+            }
+        }
+
+        impl<
+                B: Dim,
+                C: Dim,
+                const H: usize,
+                const W: usize,
+                const OH: usize,
+                const OW: usize,
+                E: Dtype,
+                D: $Kernel<E> + ZerosTensor<E>,
+                T: 'static + Tape<E, D>,
+            > $ConstTrait<OH, OW> for Tensor<(B, C, Const<H>, Const<W>), E, D, T>
+        where
+        {
+            type Output = Tensor<
+                (
+                    B,
+                    C,
+                    Const<OH>,
+                    Const<OW>,
+                ),
+                E,
+                D,
+                T,
+            >;
+
+            fn try_upscale2d(self) -> Result<Self::Output, Self::Err> {
+                let &(batch, chan, _, _) = self.shape();
+                let op = Upscale2DOp::new([batch.size(), chan.size(), H, W], [OH, OW]);
+                let (inp, mut tape) = self.split_tape();
+                let mut out = inp.device.try_zeros_like(&(
+                    batch,
+                    chan,
+                    Default::default(),
+                    Default::default(),
+                ))?;
+                inp.device.forward(op, &inp, &mut out)?;
+                let phantom_out = out.clone();
+                tape.try_alloc_grad(&inp)?;
+                tape.try_alloc_grad(&out)?;
+                tape.add_backward_op(move |grads| {
+                    let (grad_inp, grad_out) = grads.mut_and_ref(&inp, &phantom_out);
+                    inp.device
+                        .backward(op, &inp, grad_inp, &phantom_out, grad_out)
+                });
+                Ok(out.put_tape(tape))
+            }
+        }
+    };
+}
+
+upscale2d!(
+    Kernel = NearestUpscale2DKernel,
+    ConstTrait = ConstNearestUpscale2D,
+    TryTrait = TryNearestUpscale2D,
+    Meth = nearest_upscale2d,
+    TryMeth = try_nearest_upscale2d
```

Comment:
Should we support other options for how to upscale? Could accept an enum, but that'd be much more complex of an implementation

Reply:
I think having an Upscale operation could be a trait, which is then implemented on Nearest and Bilinear, I'd like at least those two.